expression/constant.go

```diff
@@ -76,15 +98,23 @@ func (c *Constant) MarshalJSON() ([]byte, error) {
 
 // Clone implements Expression interface.
 func (c *Constant) Clone() Expression {
-	if c.DeferredExpr != nil {
+	if c.DeferredExpr != nil || c.ParamMarker != nil {
 		con := *c
 		return &con
 	}
 	return c
 }
 
+var unspecifiedTp = types.NewFieldType(mysql.TypeUnspecified)
+
 // GetType implements Expression interface.
 func (c *Constant) GetType() *types.FieldType {
+	if c.ParamMarker != nil {
```

Comment:
How about:
make `c.ParamMarker.tp` to be a pointer,

```go
if c.ParamMarker != nil {
		if c.ParamMarker.tp != nil {
			return c.ParamMarker.tp
		}
		tp := types.NewFieldType(mysql.TypeUnspecified)
		types.DefaultParamTypeForValue(c.ParamMarker.GetUserVar(), tp)
		c.ParamMarker.tp = tp
		return tp
	}
```

Reply:
normally prepare an stmt `pstmt = prepare(select ?)` and execute twice,  GetUserVar should be the same type.

but client maybe first time `pstmt.SetInt(1, 1); pstmt.execute();` then `pstmt.SetString(1, "1"); pstmt.execute()`?

Reply:
Sorry, my suggestion has some problem, it is not suitable when prepare plan cache is enabled.